rubicon/objc/runtime.py

```diff
@@ -2113,3 +2128,11 @@ def __init__(self, func, restype=NOTHING, *arg_types):
 
     def wrapper(self, instance, *args):
         return self.func(*args)
+
+    def disposeHelper(self, dst):
+        Block._keep_alive_blocks_.pop(self.block.value, None)
+
+    def copyHelper(self, dst, src):
+        Block._keep_alive_blocks_.pop(self.block.value, None)
+        self.block = objc_block(dst)
+        Block._keep_alive_blocks_[self.block.value] = self
```

Comment:
I'm not sure if modifying `self` is the right approach here. When a block is copied, the original stays alive and valid, so I think `self` should continue to point to the old block. `Block._keep_alive_blocks_[dst.value] = self` should be enough - that creates an extra reference that stays even if the old block is deleted. And since the new block has the same descriptor as the old one, it also has the same dispose helper, so once the new block is deleted, its entry in the table will be removed as well.

(Actually, that might require a small modification to the dispose helper - it should remove the entry for `dst.value` from the table, not `self.block.value`, otherwise the same entry would be deleted twice.)

Reply:
Ok, I'll look at your suggestions about Python CFUNCTYPE stuff -- I'm not a Python expert so much as I am an ObjC guy.

As for your suggestion about not updating the pointer -- I think this would lead to incorrect behavior, for what we want. 

After ObjC calls block copyHelper, it no longer ever refers to that block by the old pointer in subsequent calls to us on the Python side.  Its concept of the Block is now 'that' new ptr.  

All subsequent calls refer to it by the new pointer (the one passed into dst), which is why I updated the entry.  Otherwise later when disposeHelper is called, we will get the new ptr passed to us, and we won't know what to do with it.

copyHelper is to inform Python that objc took ownership of a copy.  ObjC thereafter always refers to it by the ptr it has, the one it owns.

If we were to dispose of our own fake blocks (which we never do in the actual way we are supposed to), a second call to disposeHelper would occur.  But since we don't actually do this (do we??) it's fine.

Actually, either way it's fine -- since our table is only concerned with keeping code alive that ObjC may call into.  Python's own garbage collection stuff is fine for us.

But I digress:

So I need to update the dict in copyHelper else disposeHelper never works right.  That's the summary.

Basically the contract is as follows, as a sequence of events in time:

1. Python owns the block B right after it creates it.  If it deallocates it, it is responsible for disposing of it (we never actually end up calling block dispose on any blocks we own -- we just let them disappear into the ether -- which is fine since it's only Python data).

2. Some time later that block is passed into an objc function and is retained (basically copied via Block_copy()) -- at which time it is 'taken over' by objective C.  Python is notified of the new copy created in copyHelper.   We don't care about our own private copy -- but we do care about ObjC's copy because we need to remember the pointer to it for later when we want to kill our Python block wrapper from memory..

3. Now, as far as ObjC is concerned, there potentially are up to 2 copies of the blocks -- the one Python owns and the one it took control over.  However the one python owns is not really important and it gets auto-cleaned from memory anyway when the Block instance dies.  The one ObjC owns is of vital concern to us because we need to be informed about when it does so the code it points to (in Python) can get cleaned up.

I am pretty sure the above is accurate.  Note I'm an ObjC guy and not a Python guy (although obviously I do know Python).

I'll take a look at your other suggestions but I am pretty sure this is correct.  Let me know if you disagree.

Reply:
Oh wait -- I just re-read what you wrote.  HA!  Ok -- you're right.  I guess what you say makes tons of sense -- we keep our private pointer for our block the same as it always was.

But -- we insert into our "remember THESE" table the pointer that ObjC is concerned with.

Got it!

Makes sense! :)